src/mono/mono/mini/interp/interp.c

```diff
@@ -5472,33 +5472,45 @@ MINT_IN_CASE(MINT_BRTRUE_I8_SP) ZEROP_SP(gint64, !=); MINT_IN_BREAK;
 		}
 		MINT_IN_CASE(MINT_CONV_OVF_U8_R4) {
 			float val = LOCAL_VAR (ip [2], float);
-			if (mono_isnan (val) || mono_trunc (val) != (guint64)val)
+			double val_r8 = val;
+			const double two64  = 4294967296.0 * 4294967296.0;
+			if (val_r8 > -1.0 && val_r8 < two64)
+				LOCAL_VAR (ip [1], guint64) = (guint64)val;
+			else
 				THROW_EX (mono_get_exception_overflow (), ip);
-			LOCAL_VAR (ip [1], guint64) = (guint64)val;
 			ip += 3;
 			MINT_IN_BREAK;
 		}
 		MINT_IN_CASE(MINT_CONV_OVF_U8_R8) {
 			double val = LOCAL_VAR (ip [2], double);
-			if (mono_isnan (val) || mono_trunc (val) != (guint64)val)
+			const double two64  = 4294967296.0 * 4294967296.0;
+			if (val > -1.0 && val < two64)
+				LOCAL_VAR (ip [1], guint64) = (guint64)val;
```

Comment:
Can we make this into a function in mono-math.h?

```c
static inline gboolean
mono_try_trunc_u64 (double val, uint64_t *out)
{
	const double two64  = 4294967296.0 * 4294967296.0;
	if (val > -1.0 && val < two64) {
		*out = (uint64_t)val;
		return TRUE;
	}
	return FALSE;
}
```

And similar for the other cases?